site/content/3.12/data-science/graph-analytics.md

```diff
@@ -0,0 +1,667 @@
+```
+
+#### Community Detection
+
+Graphs based on real world networks often have a community structure.
+This means it is possible to find groups of vertices such that each vertex
+group is internally more densely connected than outside the group.
+This has many applications when you want to analyze your networks, for example
+Social networks include community groups (the origin of the term, in fact)
+based on common location, interests, occupation, etc.
+
+##### Label Propagation
+
+`POST <ENGINE_URL>/v1/labelpropagation`
+
+*Label Propagation* can be used to implement community detection on large
+graphs. The algorithm assigns a community, more precisely, a Community ID 
+(a natural number), to every vertex in the graph. 
+The idea is that each vertex should be in the community that most of
+its neighbors are in. 
+
+At first, the algorithm assigns unique initial Community IDs to the vertices. 
+The assignment is deterministic given the graph and the distribution of vertices
+on the shards, but there is no guarantee that a vertex obtains
+the same initial ID in two different runs of the algorithm, even if the graph does not change 
+(because the sharding may change). Moreover, there is no guarantee on a particular
+distribution of the initial IDs over the vertices.
+
+Then, in each iteration, a vertex sends its current Community
+ID to all its neighbor vertices. After that each vertex adopts the Community ID it
+received most frequently in the last step. 
+
+Note that, in a usual implementation of Label Propagation, if there are
+multiple most frequently received Community IDs, one is chosen randomly.
+An advantage of our implementation is that this choice is deterministic.
+This comes for the price that the choice rules are somewhat involved: 
+If a vertex obtains only one ID and the ID of the vertex from the previous step,
+its old ID, is less than the obtained ID, the old ID is kept. 
+(IDs are numbers and thus comparable to each other.) If a vertex obtains
+more than one ID, its new ID is the lowest ID among the most frequently 
+obtained IDs. (For example, if the obtained IDs are 1, 2, 2, 3, 3,
+then 2 is the new ID.) If, however, no ID arrives more than once, the new ID is
+the minimum of the lowest obtained IDs and the old ID. (For example, if the
+old ID is 5 and the obtained IDs are 3, 4, 6, then the new ID is 3. 
+If the old ID is 2, it is kept.) 
+
+If a vertex keeps its ID 20 times or more in a row, it does not send its ID.
+Vertices that did not obtain any IDs do not update their ID and do not send it.
+
+The algorithm runs until it converges, which likely never really happens on
+large graphs. Therefore you need to specify a maximum iteration bound.
+The default bound is 500 iterations, which is too large for
+common applications. 
+
+The algorithm should work best on undirected graphs. On directed
```

Comment:
The current algorithm treats every edge as undirected and thus follows edges in the forward as well as in the backwards direction.